### PR TITLE
feat: propagate EmptyRelation for more join types

### PR DESCRIPTION
**Pull request by tshauck**
## Which issue does this PR close?

Doesn't fully close, but related to #10967

## Rationale for this change

Currently the `PropagateEmptyRelation` relation only can optimize inner joins (probably the most important).

This PR updates the join handling to include simple cases for left, right, left-semi, right-semi, left-anti, and right-anti joins.

## What changes are included in this PR?

Slight refactoring of the join match statement `PropagateEmptyRelation::rewrite` then add the aforementioned joins.

## Are these changes tested?

Add tests for the added join types to confirm they're rewritten to an empty relation.

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 1]
```diff
@@ -400,6 +433,182 @@ mod tests {
         assert_together_optimized_plan_eq(plan, expected)
     }
 
+    #[test]
```
Comment (tshauck) [inline datafusion/optimizer/src/propagate_empty_relation.rs:436]: I'm realizing these could/should be slt tests?

File: datafusion/sqllogictest/test_files/subquery.slt [hunk 2]
```diff
@@ -613,10 +613,7 @@ SELECT t1_id, t1_name FROM t1 WHERE EXISTS (SELECT * FROM t2 WHERE t2_id = t1_id
 query TT
 explain SELECT t1_id, t1_name FROM t1 WHERE EXISTS (SELECT * FROM t2 WHERE t2_id = t1_id limit 0)
 ----
-logical_plan
-01)LeftSemi Join: t1.t1_id = __correlated_sq_1.t2_id
-02)--TableScan: t1 projection=[t1_id, t1_name]
-03)--EmptyRelation
+logical_plan EmptyRelation
```
Comment (tshauck) [inline datafusion/sqllogictest/test_files/subquery.slt:616]: I think the EmptyRelation in the original test is causing the entire join to be optimized out. Is there a way I can preserve this test's testing of just the subquery logic?

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 3]
```diff
@@ -94,29 +94,62 @@ impl OptimizerRule for PropagateEmptyRelation {
...
                 // For Full Join, only both sides are empty, the Join result is empty.
                 // For LeftOut/Full Join, if the right side is empty, the Join can be eliminated with a Projection with left side
                 // columns + right side columns replaced with null values.
                 // For RightOut/Full Join, if the left side is empty, the Join can be eliminated with a Projection with right side
                 // columns + left side columns replaced with null values.
                 let (left_empty, right_empty) = binary_plan_children_is_empty(&plan)?;
-                if left_empty || right_empty {
-                    return Ok(Transformed::yes(LogicalPlan::EmptyRelation(
-                        EmptyRelation {
+
```
Comment (alamb) [inline datafusion/optimizer/src/propagate_empty_relation.rs:107]: I double checked that this code faithfully implements the semantics in the comments, and I did some mental review and verification to convince myself that these semantics are correct

i wonder if @jackwener  has a moment to double check too?

**Review by alamb — APPROVED** [review]
Thank you @tshauck  -- I think this looks like a nice improvement to me

I agree it would be nice to add some slt tests for this feature as well as maybe reduce repetition in the unit tests, but I don't think they are required to merge the PR

I'll plan to leave this one open for another day or two to give others a chance to review it as well if they would like

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 4]
```diff
@@ -400,6 +433,182 @@ mod tests {
...
+                vec![Column::from_name("a".to_string())],
+            )?
+            .build()?;
+
+        let expected = "EmptyRelation";
+        assert_together_optimized_plan_eq(plan, expected)
+    }
+
+    #[test]
+    fn test_right_anti_join_empty_right_table() -> Result<()> {
```
[the remark this answers is not in the document]
Comment (tshauck) [reply datafusion/optimizer/src/propagate_empty_relation.rs]: Please see https://github.com/apache/datafusion/blob/d8bd11a6428987808f2ffe9f7beec315983c03cd/datafusion/optimizer/src/propagate_empty_relation.rs#L504-L557 (second half)

**Comment by tshauck** [comment]
Thanks for the feedback @alamb, I took some steps in [66aa31a](https://github.com/apache/datafusion/pull/10963/commits/66aa31a1d9833950db125d86b1f872e01c4952e1) to improve the tests by reducing redundancy and adding negative cases

**Review by jonahgao — APPROVED** [review]
I think these rewrites are correct.

File: datafusion/optimizer/src/test/mod.rs [hunk 5]
```diff
@@ -169,24 +184,42 @@ pub fn assert_optimized_plan_eq(
...
+pub fn assert_optimized_plan_ne_with_rules(
+    rules: Vec<Arc<dyn OptimizerRule + Send + Sync>>,
+    plan: LogicalPlan,
+    expected: &str,
+) -> Result<()> {
+    let optimized_plan = generate_optimized_plan_with_rules(rules, plan);
+    let formatted_plan = format!("{optimized_plan:?}");
+    assert_ne!(formatted_plan, expected);
+    Ok(())
+}
```
Comment (jackwener) [inline datafusion/optimizer/src/test/mod.rs:221]: we can merge them into one function and use a param to determine `assert_ne` or `assert_eq`

Like
```rust 
pub fn assert_optimized_plan_eq_with_rules(
    rules: Vec<Arc<dyn OptimizerRule + Send + Sync>>,
    plan: LogicalPlan,
    expected: &str,
    eq: bool,
) -> Result<()> {
    let optimized_plan = generate_optimized_plan_with_rules(rules, plan);
    let formatted_plan = format!("{optimized_plan:?}");
    if eq {
        assert_eq!(formatted_plan, expected);
    } else {
        assert_ne!(formatted_plan, expected);
    }
    Ok(())
}
```

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 6]
```diff
@@ -94,29 +94,62 @@ impl OptimizerRule for PropagateEmptyRelation {
...
+                            produce_one_row: false,
+                            schema: join.schema.clone(),
+                        }),
+                    )),
+                    JoinType::RightAnti if right_empty => Ok(Transformed::yes(
+                        LogicalPlan::EmptyRelation(EmptyRelation {
+                            produce_one_row: false,
+                            schema: join.schema.clone(),
+                        }),
+                    )),
```
Comment (jackwener) [inline datafusion/optimizer/src/propagate_empty_relation.rs:150]: Why don't we delete above comment about `anti join`, it already handle anti-join here

**Review by jackwener — COMMENTED** [review]
Except for some small details, I have no problem with this PR. 
In addition, I hope to see more `slt` tests.

Thanks @tshauck

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 7]
```diff
@@ -94,29 +94,62 @@ impl OptimizerRule for PropagateEmptyRelation {
                 Ok(Transformed::no(LogicalPlan::CrossJoin(join.clone())))
             }
 
-            LogicalPlan::Join(ref join) if join.join_type == Inner => {
+            LogicalPlan::Join(ref join) => {
                 // TODO: For Join, more join type need to be careful:
```
Comment (jackwener) [inline datafusion/optimizer/src/propagate_empty_relation.rs:98]: Does these comment need to remain?

File: datafusion/optimizer/src/propagate_empty_relation.rs (hunk 6 above)
Comment (tshauck) [reply datafusion/optimizer/src/propagate_empty_relation.rs:150]: This only handles the simple case for anti join where the respective left or right side is empty.

File: datafusion/optimizer/src/propagate_empty_relation.rs (hunk 7 above)
Comment (tshauck) [reply datafusion/optimizer/src/propagate_empty_relation.rs:98]: I removed only the ones I added, if we merge this I'll update the associated ticket.

File: datafusion/optimizer/src/test/mod.rs (hunk 5 above)
Comment (tshauck) [reply datafusion/optimizer/src/test/mod.rs]: Updated in 401f2e22b1c9aac23bf4e786847b7741141d40de

File: datafusion/optimizer/src/propagate_empty_relation.rs (hunk 6 above)
Comment (jackwener) [reply datafusion/optimizer/src/propagate_empty_relation.rs:150]: Thanks, I notice it.

**Review by alamb — APPROVED** [review]
Looks great to me -- thank  you @tshauck @jonahgao and @jackwener

**Pull request merged**